File: backend-microservices/backend-go/audit.go
package main

import (
    "database/sql"
    "encoding/json"
    _ "github.com/lib/pq"
    "time"
)

func AuditLog(db *sql.DB, action, entity, entityID, userID, requestIP string, details map[string]any) error {
    jsonDetails, _ := json.Marshal(details)

    _, err := db.Exec(`
        INSERT INTO audit_log (timestamp, service, user_id, action, entity, entity_id, request_ip, details)
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
    `,
        time.Now().UTC(),
        "backend-go",
        userID,
        action,
        entity,
        entityID,
        requestIP,
        jsonDetails,
    )

    return err
}

File: backend-microservices/backend-python/audit.py
import psycopg2
import os
import json
from datetime import datetime


def audit_log(action, entity=None, entity_id=None, user_id=None, request_ip=None, details=None):
    conn = psycopg2.connect(os.getenv("DATABASE_URL"))
    cur = conn.cursor()

    cur.execute("""
        INSERT INTO audit_log (timestamp, service, user_id, action, entity, entity_id, request_ip, details)
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
    """, (
        datetime.utcnow(),
        "backend-python",
        user_id,
        action,
        entity,
        entity_id,
        request_ip,
        json.dumps(details) if details else None
    ))

    conn.commit()
    cur.close()
    conn.close()
